### PR TITLE
metashift

### PR DESCRIPTION
**Pull request by SamuelMathieu-code**
Added metashift dataset with metashift.py

### CODE REVIEW

File: continuum/datasets/metashift.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+          2022
+          arXiv:2202.06523v1
+
+    :param data_path: The folder path containing the data.
+    :param train : 
+    :param download: If true, will download images and the dictionnary 
+           containing classes and contexts if not already in data_path folder.
+    :param train_image_ids: Images ids to use.
+    :param class_names : classes to consider (default = None --> all classes)
+    :param random_context : If true ; only one occurence of each image in a
```
Comment (TLESORT) [inline continuum/datasets/metashift.py:29]: you can rename that "unique_occurrence" it would be clearer

File: continuum/datasets/metashift.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    :param data_path: The folder path containing the data.
+    :param train : 
+    :param download: If true, will download images and the dictionnary 
+           containing classes and contexts if not already in data_path folder.
+    :param train_image_ids: Images ids to use.
+    :param class_names : classes to consider (default = None --> all classes)
+    :param random_context : If true ; only one occurence of each image in a 
+           random choosen class&context combination. If false, all valid contexts are considered, 
+           duplicates of ids will be found in x.
+    :param random_seed : set seed (relevant ponly if random context is True)
```
Comment (TLESORT) [inline continuum/datasets/metashift.py:32]: *only

File: continuum/datasets/metashift.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    def __init__(
+            self,
+            data_path:str,
+            train:bool = True,
+            download:bool = True,
+            class_names:Union[Iterable[str], None] = None, #Only get images corresponding to specific classe(s)
+            train_image_ids:Union[Iterable[str], None] = None, #Only get specific ids for training. (if None : all)
+            random_contexts:bool = False, #If true, the images will be assigned random class(context) combination among all valid combinations.
+            random_seed:int = 42,
+            nb_tasks:int = 0
```
Comment (TLESORT) [inline continuum/datasets/metashift.py:47]: A supplementary option that say if all classes should appear in all tasks would be nice.
with a name like 'strict_domain_inc'.

File: docs/tutorials/scenarios_suites/MetaShift.rst [hunk 4]
```diff
@@ -0,0 +1,72 @@
+MetaShift
+-----------------
+
+MetaShift can be described as a dataset of dataset. It divides the original Visual Genome dataset into classes and contexts in order to capture distribution shifts.
+See paper : https://arxiv.org/abs/2202.06523
+
+The MetaShift class is built with ContinualScenario in mind. But it also lets the user select specific classes or training ids. 
+As many images can be found in more than one class(context) configuration, the user can specify that all images should aprear in exactly one configuration.
```
Comment (TLESORT) [inline docs/tutorials/scenarios_suites/MetaShift.rst:8]: peut etre qu'une autre facon de formuler ca serait:

In continuum, we use the context information to build various types of scenarios, notably domain incremental scenario.
The user can select a subset of classes to include in the scenario with `class_names` argument and if all classes should appear in all tasks with `???` arguments.
Images can be associated with several context, then the user can decide if one image appear only once or can appear several time.
If the image should appear only once, then the context chosen will be selected randomly.

File: docs/tutorials/scenarios_suites/MetaShift.rst [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+To download the Visual Genome dataset in advance :
+
+.. code-block:: bash
+
+    wget -c https://nlp.stanford.edu/data/gqa/images.zip
+    unzip images.zip -d images
+
+
+Images will have to be stored this way :
```
Comment (TLESORT) [inline docs/tutorials/scenarios_suites/MetaShift.rst:21]: si c'est pas une operation manuel, tu peux retirer cette partie sur la facon dont les images sont stockees

File: docs/tutorials/scenarios_suites/MetaShift.rst [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+.. code-block:: python
+
+    from continuum.datasets import MetaShift
+    from continuum.scenarios import ContinualScenario
+
+    data = MetaShift(datafolder, class_names = ["cat", "dog"])
+    scenario = ContinualScenario(data)
+
+
+- Specific ids:
```
Comment (TLESORT) [inline docs/tutorials/scenarios_suites/MetaShift.rst:53]: C'est quoi les ids?

File: docs/tutorials/scenarios_suites/MetaShift.rst [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+
+- Get a unique task for each image:
+
+.. code-block:: python
+
+    from continuum.datasets import MetaShift
+    from continuum.scenarios import ContinualScenario
+
+    data = MetaShift(datafolder, random_contexts=Ture)
```
Comment (TLESORT) [inline docs/tutorials/scenarios_suites/MetaShift.rst:71]: *True

File: continuum/datasets/metashift.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    """
+    data_url = "https://nlp.stanford.edu/data/gqa/images.zip"
+    pickle_url = "https://github.com/Weixin-Liang/MetaShift/blob/main/dataset/meta_data/full-candidate-subsets.pkl?raw=true"
+
+    def __init__(
+            self,
+            data_path:str,
+            train:bool = True,
+            download:bool = True,
+            class_names:Union[Iterable[str], None] = None, #Only get images corresponding to specific classe(s)
```
Comment (TLESORT) [inline continuum/datasets/metashift.py:45]: maybe class number would be better, what are the class names?

File: continuum/datasets/metashift.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+        n = np.max(t) + 1
+
+        if (self.nb_tasks > 0) and (n > self.nb_tasks): # to be tested
+            scale = lambda x : x % self.nb_tasks # scale task ids to number of tasks
+            t = scale(t)
+
+        return x, y, t
+
+def _select_unique_occurence(x, y, t, rand_seed): 
+    #choose a random context for each class(context) combiation for each image amoung available.
```
Comment (TLESORT) [inline continuum/datasets/metashift.py:147]: do you want to say that ?
`# choose an unique context for each data point (randomly)`

File: continuum/datasets/metashift.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+            x, y, t = _strict_domain_tasks(x, y, t, len(self.class_names))
+            t = np.unique(t, return_inverse=True)[1]
+        else:
+            t = np.unique(t, return_inverse = True)[1]
+        
+        n = np.max(t) + 1
+
+        if (self.nb_tasks > 0) and (n > self.nb_tasks): # to be tested
+            scale = lambda x : x % self.nb_tasks # scale task ids to number of tasks
+            t = scale(t)
```
Comment (TLESORT) [inline continuum/datasets/metashift.py:142]: pourquoi pas juste:
`t = t % self.nb_tasks` ?

File: continuum/datasets/metashift.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+    for i in range(len(vals)):
+        final_idx_list.append(randint(idx_start[i], idx_start[i]+count[i]-1))
+        
+    x2 = sorted_x[final_idx_list]
+    y2 = sorted_y[final_idx_list]
+    t2 = sorted_t[final_idx_list]
+
+    return x2, y2, t2
+
+def _strict_domain_tasks(x, y, t, nb_classes):
```
Comment (TLESORT) [inline continuum/datasets/metashift.py:168]: ca peut pas marcher ca je pense.
Il faut remplace `nb_classes` par `classes_ids` imo

File: docs/tutorials/scenarios_suites/MetaShift.rst [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+Visual Genome:
+##############
+
+To download the Visual Genome dataset in advance :
+
+.. code-block:: bash
+
+    wget -c https://nlp.stanford.edu/data/gqa/images.zip
+    unzip images.zip -d images
```
Comment (TLESORT) [inline docs/tutorials/scenarios_suites/MetaShift.rst:22]: si le dataset se DL automatiquement, est ce que c'est utile?

File: docs/tutorials/scenarios_suites/MetaShift.rst [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
```
Comment (TLESORT) [inline docs/tutorials/scenarios_suites/MetaShift.rst:23]: Adding some images for scenarios illustration would be nice

File: docs/tutorials/scenarios_suites/MetaShift.rst [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+Select specific classes to appear in the dataset with the argument :code:`class_names`.
+Then specify if all classes should apprear in all tasks with the argument :code:`strict_domain_inc`. If True, only contexts found in all classes will be kept.
+
+.. code-block:: python
+
+    from continuum.datasets import MetaShift
+    from continuum.scenarios import ContinualScenario
+
+    data = MetaShift(datafolder, class_names = ["cat", "dog"], strict_domain_inc = True)
```
Comment (TLESORT) [inline docs/tutorials/scenarios_suites/MetaShift.rst:49]: providing somewhere the list of classes would be usefull

File: docs/tutorials/scenarios_suites/MetaShift.rst (hunk 12 above)
Comment (TLESORT) [reply docs/tutorials/scenarios_suites/MetaShift.rst:22]: est ce qu'il est sensé se DL automatiquement? quand j'essaye ca ne marche pas je pense

Comment (TLESORT) [reply docs/tutorials/scenarios_suites/MetaShift.rst:22]: si ca peut pas se DL auto, il faut mettre le datasets dans `*data_dir*/MetaShift`

Comment (SamuelMathieu-code) [reply docs/tutorials/scenarios_suites/MetaShift.rst:22]: J'ai enlevé le dl auto pcq c'était quand même 20GB. Ça me paraissait beaucoup. Mais si tu veux je le remets. :)

File: continuum/datasets/metashift.py [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+        if os.path.exists(os.path.join(self.data_path, "full-candidate-subsets.pkl")):
+            print("Classes and contexts already downloaded")
+        else:
+            file = download.download(self.pickle_url, self.data_path)
+            os.rename(file, os.path.join(self.data_path, "full-candidate-subsets.pkl"))
+            print("\nClasses and contexts downloaded")
+        
+        # Visual Genome targets
+        if os.path.exists(os.path.join(self.data_path, "sceneGraphs")):
+            print("Visual Genome targets already downloaded")
```
Comment (TLESORT) [inline continuum/datasets/metashift.py:91]:
```
print("Metashift Targets already downloaded")
```
maybe the user do not necessarily know that it is using visual genome dataset

File: continuum/datasets/metashift.py [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+        else:
+            file = download.download(self.targets_url, self.data_path)
+            path = os.path.join(self.data_path, "sceneGraphs")
+            if not os.path.exists(path) : os.mkdir(path)
+            with zipfile.ZipFile(file, 'r') as zip_file:
+                zip_file.extractall(path)
+            print("\nVisual Genome targets downloaded")
```
Comment (TLESORT) [inline continuum/datasets/metashift.py:98]: same as previous comment

File: docs/tutorials/scenarios_suites/MetaShift.rst [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+
+    from continuum.datasets import MetaShift
+    data = MetaShift(datafolder, visual_genome_path = "some/path/images")
+
+ContinualScenario:
+##################
+
+Beware that the default configuration will download the full Visual Genome dataset if it is not already present in a folder named "MetaShift" (20GB).
+
+Tasks are defined by the context in whcich  the objects appear. In the following example, dogs and cats with water on the right and with computers on the left.
```
Comment (TLESORT) [inline docs/tutorials/scenarios_suites/MetaShift.rst:37]: which

File: docs/tutorials/scenarios_suites/MetaShift.rst [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+
+    from continuum.datasets import MetaShift
+    from continuum.scenarios import ContinualScenario
+    from continuum.datasets.metashift import get_all_classes_contexts
+
+    all_classes, all_contexts = get_all_classes_contexts(datafolder) 
+    # Use this function to retrieve all metashift classes and contexts.
+
+    data = MetaShift(datafolder, class_names = ["cat", "dog"], strict_domain_inc = True)
+    scenario = ContinualScenario(data)
```
Comment (TLESORT) [inline docs/tutorials/scenarios_suites/MetaShift.rst:70]: Can you add a comment here, to say how many tasks will there be here?

**Pull request merged**